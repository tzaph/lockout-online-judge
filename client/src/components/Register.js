import React, { useState } from "react";
import { Alert } from "react-bootstrap";
import { useAuth } from "../contexts/AuthContext";
import { Link, useNavigate } from "react-router-dom";

export function isEmailValid(email) {
  const regex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return regex.test(email);
}

export function isPasswordMatch(pass, confirmPass) {
  if (pass !== confirmPass) {
    return false;
  }
  return true;
}

export function isPasswordLong(pass) {
  if (pass.length < 6) {
    return false;
  }
  return true;
}

const Register = () => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");
  const register = useAuth()?.register;
  const addUserInformationToDatabase = useAuth()?.addUserInformationToDatabase;
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);
  const navigate = useNavigate();

  const handleSubmit = async (e) => {
    e.preventDefault();
    if (!isEmailValid(email)) {
      return setError("Please enter a valid email!");
    }
    if (!isPasswordMatch(password, confirmPassword)) {
      return setError("Passwords do not match!");
    }
    if (!isPasswordLong(password)) {
      return setError("Password must be at least 6 characters long!");
    }
    try {
      await register(email, password);
      await addUserInformationToDatabase();
      setLoading(true);
      setError("");
      navigate("/");
    } catch {
      setError("Failed to create a new account");
    }
    setLoading(false);
  };

  return (
    <div className="auth-form-container">
      <h2>Register</h2>
      {error && <Alert variant="danger">{error}</Alert>}
      <form className="register-form" onSubmit={handleSubmit}>
        <label for="email">Email</label>
        <input
          value={email}
          onChange={(e) => setEmail(e.target.value)}
          type="email"
          placeholder="youremail@gmail.com"
          id="email"
          name="email"
        />
        <label for="password">Password</label>
        <input
          value={password}
          onChange={(e) => setPassword(e.target.value)}
          type="password"
          placeholder="********"
          id="password"
          name="password"
        />
        <label for="confirm-password">Confirm Password</label>
        <input
          value={confirmPassword}
          onChange={(e) => setConfirmPassword(e.target.value)}
          type="password"
          placeholder="********"
          id="confirmPassword"
          name="confirmPassword"
        />
        <button disabled={loading} type="submit">
          Register
        </button>
      </form>
      <div className="link-button">
        Already have an account? <Link to="/login">Login</Link>
      </div>
    </div>
  );
};

export default Register;
