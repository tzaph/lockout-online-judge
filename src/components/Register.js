import React, { useState } from 'react';
import { Alert } from 'react-bootstrap';
import { useAuth } from '../contexts/AuthContext';
import { Link, useNavigate } from 'react-router-dom';

const Register = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const { register } = useAuth();
  const [error, setError] = useState('');
  const [loading, setLoading] = useState(false);
  const navigate = useNavigate();
  
  const handleSubmit = async(e) => {
    e.preventDefault();
    try {
      register(email, password);
      setLoading(true);
      setError('');
      navigate('/');
    } catch {
      setError('Failed to create a new account');
    }
    setLoading(false);
  }

  return (
    <div className="auth-form-container">
      <h2>Register</h2>
      {error && <Alert variant='danger'>{error}</Alert>}
      <form className="register-form" onSubmit={handleSubmit}>
        <label for="email">Email</label>
        <input
          value={email}
          onChange={(e) => setEmail(e.target.value)}
          type="email"
          placeholder="youremail@gmail.com"
          id="email"
          name="email"
        />
        <label for="password">Password</label>
        <input
          value={password}
          onChange={(e) => setPassword(e.target.value)}
          type="password"
          placeholder="********"
          id="password"
          name="password"
        />
        <button disabled={loading} type="submit">Register</button>
      </form>
      <div className="link-button">
        Already have an account? <Link to="/login">Login</Link>
      </div>
    </div>
  )
}

export default Register;