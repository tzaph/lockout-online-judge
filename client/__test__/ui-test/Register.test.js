import React from "react";
import renderer from "react-test-renderer";
import Register from "../../src/components/Register";
import { BrowserRouter as Router } from "react-router-dom";

describe("<Register />", () => {
  it("Renders correctly", () => {
    const tree = renderer
      .create(
        <Router>
          <Register />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
