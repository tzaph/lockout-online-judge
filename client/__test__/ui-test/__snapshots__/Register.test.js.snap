// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> Renders correctly 1`] = `
<div
  className="auth-form-container"
>
  <h2>
    Register
  </h2>
  <form
    className="register-form"
    onSubmit={[Function]}
  >
    <label
      for="email"
    >
      Email
    </label>
    <input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="youremail@gmail.com"
      type="email"
      value=""
    />
    <label
      for="password"
    >
      Password
    </label>
    <input
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="********"
      type="password"
      value=""
    />
    <label
      for="confirm-password"
    >
      Confirm Password
    </label>
    <input
      id="confirmPassword"
      name="confirmPassword"
      onChange={[Function]}
      placeholder="********"
      type="password"
      value=""
    />
    <button
      disabled={false}
      type="submit"
    >
      Register
    </button>
  </form>
  <div
    className="link-button"
  >
    Already have an account? 
    <a
      href="/login"
      onClick={[Function]}
    >
      Login
    </a>
  </div>
</div>
`;
